Improve handling for hybrid svelte packages in automatic dependency optimization

### Describe the problem

Some svelte packages are hybrids and could contain dependencies on node-only packages.

These packages then can fail in vite's optimizer with an error like 
`
Failed to resolve entry for package "foo > bar". The package may have incorrect main/module/exports specified in its package.json
`

Currently known hybrids:
* `@roxi/routify` - error caused by `@types/node` (only `types` entry), `cheap-watch`, `rollup-pluginutils` (export esm and used node builtin modules)
* `@carbon/charts-svelte` - error caused by `@carbon/telemetry` (only `bin` entry)

### Describe the proposed solution

* detect dependencies that would fail in vite's optimizer and don't reinclude them.
* detect dependencies that are not used in svelte components and don't reinclude them

### Alternatives considered

* Improve vite's optimizer to allow deep scanning
* Hardcode list of known hybrids
* Keep the status quo where people have to opt-out of the automatic behavior
  * either by adding the dependencies that fail in vite's optimizer to `optimizeDeps.exclude` in vite config
  * or by adding the declaring package to vite-plugin-svelte's [disableDependencyReinclusion](https://github.com/sveltejs/vite-plugin-svelte/blob/main/docs/config.md#disabledependencyreinclusion) option.
* Change automatic dependency handling to be opt-in in vite-plugin-svelte  

### Importance

would make my life easier

@benmccann I've updated the description with some info on why the packages fail.

For `@types/node` and `@carbon/telemetry`, their only entrypoints are `types` and `bin`, and that makes sense as they're only ever used in those contexts (type definition and command). The best heuristic we could use for these packages are by looking up their `package.json` and checks if they have main/module/export (other custom main fields won't be checked though but that's likely a packaging issue within the library itself.

For `cheap-watch` and `rollup-pluginutil`, this is really weird. These are node-only deps, but Routify also has other node-only deps, so why these two libraries only specifically? Turns out these two libraries export ESM code, which tripped Vite's pre-bundling to this error:

```
node_modules/cheap-watch/dist/CheapWatch.esm.js:2:9: error: No matching export in "browser-external:fs" for import "watch"
```

The other node-only deps only exported CJS code and Vite worked fine with them (Likely because the prebundling `needsInterop` metadata helped caught these?). But this whole kerfuffle is ironic in general since we would always encourage libraries to export ESM code.

---

One solution I had in mind was to implement our own scanner and detect used dependencies ourselves, instead of the current implementation that puts all the library's `dependencies` into optimization. It could be tricky though with a lot of unknowns ahead.

Otherwise, ideally we could fix by letting Vite do all the hard work of nested optimization by default, addressing `Improve vite's optimizer to allow deep scanning`, but I haven't looked into how feasible that is yet.

--- a/.changeset/tricky-clouds-grab.md
+++ b/.changeset/tricky-clouds-grab.md
@@ -1,0 +1,5 @@
+---
+'@sveltejs/vite-plugin-svelte': patch
+---
+
+Only optimize nested cjs dependencies

--- a/packages/e2e-tests/package-json-svelte-field/__tests__/package-json-svelte-field.spec.ts
+++ b/packages/e2e-tests/package-json-svelte-field/__tests__/package-json-svelte-field.spec.ts
@@ -7,14 +7,13 @@ test('should render component imported via svelte field in package.json', async 
 });
 
 if (!isBuild) {
-	test('should optimize nested deps of excluded svelte deps', () => {
+	test('should optimize nested cjs deps of excluded svelte deps', () => {
 		const vitePrebundleMetadata = path.resolve(__dirname, '../node_modules/.vite/_metadata.json');
 		const metadataFile = fs.readFileSync(vitePrebundleMetadata, 'utf8');
 		const metadata = JSON.parse(metadataFile);
 		const optimizedPaths = Object.keys(metadata.optimized);
 		expect(optimizedPaths).not.toContain('e2e-test-dep-svelte-nested');
 		expect(optimizedPaths).not.toContain('e2e-test-dep-svelte-simple');
-		expect(optimizedPaths).toContain('e2e-test-dep-svelte-nested > e2e-test-dep-cjs-and-esm');
 		expect(optimizedPaths).toContain(
 			'e2e-test-dep-svelte-nested > e2e-test-dep-svelte-simple > e2e-test-dep-cjs-only'
 		);

--- a/packages/e2e-tests/package-json-svelte-field/vite.config.js
+++ b/packages/e2e-tests/package-json-svelte-field/vite.config.js
@@ -3,11 +3,7 @@ const { defineConfig } = require('vite');
 
 module.exports = defineConfig(({ command, mode }) => {
 	return {
-		plugins: [
-			svelte({
-				disableDependencyReinclusion: ['e2e-test-dep-svelte-hybrid']
-			})
-		],
+		plugins: [svelte()],
 		build: {
 			// make build faster by skipping transforms and minification
 			target: 'esnext',

--- a/packages/vite-plugin-svelte/src/index.ts
+++ b/packages/vite-plugin-svelte/src/index.ts
@@ -52,7 +52,7 @@ export function svelte(inlineOptions?: Partial<Options>): Plugin {
 			}
 			options = await resolveOptions(inlineOptions, config, configEnv);
 			// extra vite config
-			const extraViteConfig = buildExtraViteConfig(options, config);
+			const extraViteConfig = buildExtraViteConfig(options, config, configEnv);
 			log.debug('additional vite config', extraViteConfig);
 			return extraViteConfig as Partial<UserConfig>;
 		},

--- a/packages/vite-plugin-svelte/src/utils/dependencies.ts
+++ b/packages/vite-plugin-svelte/src/utils/dependencies.ts
@@ -2,7 +2,6 @@ import { log } from './log';
 import path from 'path';
 import fs from 'fs';
 import { createRequire } from 'module';
-import { SVELTE_RESOLVE_MAIN_FIELDS } from './constants';
 
 export function findRootSvelteDependencies(root: string, cwdFallback = true): SvelteDependency[] {
 	log.debug(`findSvelteDependencies: searching svelte dependencies in ${root}`);
@@ -46,14 +45,6 @@ function getSvelteDependencies(
 	for (const { pkg, dir } of resolvedDeps) {
 		result.push({ name: pkg.name, pkg, dir, path });
 		if (pkg.dependencies) {
-			// removed unoptimizable dependencies (hacky)
-			for (const dep in pkg.dependencies) {
-				const localRequire = createRequire(`${dir}/package.json`);
-				const depData = resolveDependencyData(dep, localRequire);
-				if (depData && !isDepOptimizable(depData.pkg)) {
-					delete pkg.dependencies[dep];
-				}
-			}
 			let dependencyNames = Object.keys(pkg.dependencies);
 			const circular = dependencyNames.filter((name) => path.includes(name));
 			if (circular.length > 0) {
@@ -171,8 +162,13 @@ function is_common_without_svelte_field(dependency: string): boolean {
 	);
 }
 
-function isDepOptimizable(pkg: Pkg) {
-	return [...SVELTE_RESOLVE_MAIN_FIELDS, 'main'].some((field) => !!pkg[field]);
+export function needsOptimization(dep: SvelteDependency, depOfDep: string) {
+	const localRequire = createRequire(`${dep.dir}/package.json`);
+	const depData = resolveDependencyData(depOfDep, localRequire);
+	if (!depData) return false;
+	const pkg = depData.pkg;
+	// only optimize if is cjs, using the below as heuristic
+	return pkg.main && !pkg.module && !pkg.exports;
 }
 
 interface DependencyData {

--- a/packages/vite-plugin-svelte/src/utils/dependencies.ts
+++ b/packages/vite-plugin-svelte/src/utils/dependencies.ts
@@ -162,9 +162,8 @@ function is_common_without_svelte_field(dependency: string): boolean {
 	);
 }
 
-export function needsOptimization(dep: SvelteDependency, depOfDep: string) {
-	const localRequire = createRequire(`${dep.dir}/package.json`);
-	const depData = resolveDependencyData(depOfDep, localRequire);
+export function needsOptimization(dep: string, localRequire: NodeRequire): boolean {
+	const depData = resolveDependencyData(dep, localRequire);
 	if (!depData) return false;
 	const pkg = depData.pkg;
 	// only optimize if is cjs, using the below as heuristic

--- a/packages/vite-plugin-svelte/src/utils/dependencies.ts
+++ b/packages/vite-plugin-svelte/src/utils/dependencies.ts
@@ -2,6 +2,7 @@ import { log } from './log';
 import path from 'path';
 import fs from 'fs';
 import { createRequire } from 'module';
+import { SVELTE_RESOLVE_MAIN_FIELDS } from './constants';
 
 export function findRootSvelteDependencies(root: string, cwdFallback = true): SvelteDependency[] {
 	log.debug(`findSvelteDependencies: searching svelte dependencies in ${root}`);
@@ -39,12 +40,20 @@ function getSvelteDependencies(
 	const result = [];
 	const localRequire = createRequire(`${pkgDir}/package.json`);
 	const resolvedDeps = deps
-		.map((dep) => resolveSvelteDependency(dep, localRequire))
-		.filter(Boolean);
+		.map((dep) => resolveDependencyData(dep, localRequire))
+		.filter((depData) => !!depData && isSvelte(depData.pkg)) as DependencyData[];
 	// @ts-ignore
 	for (const { pkg, dir } of resolvedDeps) {
 		result.push({ name: pkg.name, pkg, dir, path });
 		if (pkg.dependencies) {
+			// removed unoptimizable dependencies (hacky)
+			for (const dep in pkg.dependencies) {
+				const localRequire = createRequire(`${dir}/package.json`);
+				const depData = resolveDependencyData(dep, localRequire);
+				if (depData && !isDepOptimizable(depData.pkg)) {
+					delete pkg.dependencies[dep];
+				}
+			}
 			let dependencyNames = Object.keys(pkg.dependencies);
 			const circular = dependencyNames.filter((name) => path.includes(name));
 			if (circular.length > 0) {
@@ -64,16 +73,10 @@ function getSvelteDependencies(
 	return result;
 }
 
-function resolveSvelteDependency(
-	dep: string,
-	localRequire: NodeRequire
-): { dir: string; pkg: Pkg } | void {
+function resolveDependencyData(dep: string, localRequire: NodeRequire): DependencyData | void {
 	try {
 		const pkgJson = `${dep}/package.json`;
 		const pkg = localRequire(pkgJson);
-		if (!isSvelte(pkg)) {
-			return;
-		}
 		const dir = path.dirname(localRequire.resolve(pkgJson));
 		return { dir, pkg };
 	} catch (e) {
@@ -84,12 +87,6 @@ function resolveSvelteDependency(
 			while (dir) {
 				const pkg = parsePkg(dir, true);
 				if (pkg && pkg.name === dep) {
-					if (!isSvelte(pkg)) {
-						return;
-					}
-					log.warn.once(
-						`package.json of ${dep} has a "svelte" field but does not include itself in exports field. Please ask package maintainer to update`
-					);
 					return { dir, pkg };
 				}
 				const parent = path.dirname(dir);
@@ -172,6 +169,15 @@ function is_common_without_svelte_field(dependency: string): boolean {
 					: dependency.substring(dependency.lastIndexOf('/') + 1).startsWith(prefix) // check prefix omitting @scope/
 		)
 	);
+}
+
+function isDepOptimizable(pkg: Pkg) {
+	return [...SVELTE_RESOLVE_MAIN_FIELDS, 'main'].some((field) => !!pkg[field]);
+}
+
+interface DependencyData {
+	dir: string;
+	pkg: Pkg;
 }
 
 export interface SvelteDependency {

--- a/packages/vite-plugin-svelte/src/utils/dependencies.ts
+++ b/packages/vite-plugin-svelte/src/utils/dependencies.ts
@@ -167,6 +167,7 @@ export function needsOptimization(dep: string, localRequire: NodeRequire): boole
 	if (!depData) return false;
 	const pkg = depData.pkg;
 	// only optimize if is cjs, using the below as heuristic
+	// see https://github.com/sveltejs/vite-plugin-svelte/issues/162
 	return pkg.main && !pkg.module && !pkg.exports;
 }
 

--- a/packages/vite-plugin-svelte/src/utils/options.ts
+++ b/packages/vite-plugin-svelte/src/utils/options.ts
@@ -15,6 +15,7 @@ import {
 import path from 'path';
 import { findRootSvelteDependencies, needsOptimization, SvelteDependency } from './dependencies';
 import { DepOptimizationOptions } from 'vite/src/node/optimizer/index';
+import { createRequire } from 'module';
 
 const knownOptions = new Set([
 	'configFile',
@@ -253,11 +254,12 @@ function buildOptimizeDepsForSvelte(
 		}
 		const transitiveDepsToInclude = svelteDeps
 			.filter((dep) => !disabledReinclusions.includes(dep.name) && isExcluded(dep.name))
-			.flatMap((dep) =>
-				Object.keys(dep.pkg.dependencies || {})
-					.filter((depOfDep) => !isExcluded(depOfDep) && needsOptimization(dep, depOfDep))
-					.map((depOfDep) => dep.path.concat(dep.name, depOfDep).join(' > '))
-			);
+			.flatMap((dep) => {
+				const localRequire = createRequire(`${dep.dir}/package.json`);
+				return Object.keys(dep.pkg.dependencies || {})
+					.filter((depOfDep) => !isExcluded(depOfDep) && needsOptimization(depOfDep, localRequire))
+					.map((depOfDep) => dep.path.concat(dep.name, depOfDep).join(' > '));
+			});
 		log.debug(
 			`reincluding transitive dependencies of excluded svelte dependencies`,
 			transitiveDepsToInclude

--- a/packages/vite-plugin-svelte/src/utils/options.ts
+++ b/packages/vite-plugin-svelte/src/utils/options.ts
@@ -1,4 +1,5 @@
 /* eslint-disable no-unused-vars */
+import { builtinModules } from 'module';
 import { ConfigEnv, UserConfig, ViteDevServer, normalizePath } from 'vite';
 import { log } from './log';
 import { loadSvelteConfig } from './load-svelte-config';
@@ -265,7 +266,28 @@ function buildOptimizeDepsForSvelte(
 		include.push(...transitiveDepsToInclude);
 	}
 
-	return { include, exclude };
+	const esbuildOptions: DepOptimizationOptions['esbuildOptions'] = {
+		plugins: [
+			{
+				name: 'vite-plugin-svelte:external-node-builtin',
+				setup(build) {
+					// externalize all node builtins during prebundling
+					// https://github.com/vitejs/vite/blob/f4f0100649220453d961b6c66531c58026885680/packages/vite/src/node/plugins/resolve.ts#L224-L241
+					build.onResolve(
+						{ filter: new RegExp(`^(${builtinModules.join('|')})$`) },
+						({ path: id }) => {
+							return {
+								path: id,
+								external: true
+							};
+						}
+					);
+				}
+			}
+		]
+	};
+
+	return { include, exclude, esbuildOptions };
 }
 
 function buildSSROptionsForSvelte(

--- a/packages/vite-plugin-svelte/src/utils/options.ts
+++ b/packages/vite-plugin-svelte/src/utils/options.ts
@@ -181,12 +181,12 @@ function resolveViteRoot(viteConfig: UserConfig): string | undefined {
 
 export function buildExtraViteConfig(
 	options: ResolvedOptions,
-	config: UserConfig
+	config: UserConfig,
+	configEnv: ConfigEnv
 ): Partial<UserConfig> {
 	// extra handling for svelte dependencies in the project
 	const svelteDeps = findRootSvelteDependencies(options.root);
 	const extraViteConfig: Partial<UserConfig> = {
-		optimizeDeps: buildOptimizeDepsForSvelte(svelteDeps, options, config.optimizeDeps),
 		resolve: {
 			mainFields: [...SVELTE_RESOLVE_MAIN_FIELDS],
 			dedupe: [...SVELTE_IMPORTS, ...SVELTE_HMR_IMPORTS]
@@ -196,6 +196,14 @@ export function buildExtraViteConfig(
 		// @ts-ignore
 		// knownJsSrcExtensions: options.extensions
 	};
+
+	if (configEnv.command === 'serve') {
+		extraViteConfig.optimizeDeps = buildOptimizeDepsForSvelte(
+			svelteDeps,
+			options,
+			config.optimizeDeps
+		);
+	}
 
 	// @ts-ignore
 	extraViteConfig.ssr = buildSSROptionsForSvelte(svelteDeps, options, config);

--- a/packages/vite-plugin-svelte/src/utils/options.ts
+++ b/packages/vite-plugin-svelte/src/utils/options.ts
@@ -1,5 +1,4 @@
 /* eslint-disable no-unused-vars */
-import { builtinModules } from 'module';
 import { ConfigEnv, UserConfig, ViteDevServer, normalizePath } from 'vite';
 import { log } from './log';
 import { loadSvelteConfig } from './load-svelte-config';
@@ -14,7 +13,7 @@ import {
 	// eslint-disable-next-line node/no-missing-import
 } from 'svelte/types/compiler/preprocess';
 import path from 'path';
-import { findRootSvelteDependencies, SvelteDependency } from './dependencies';
+import { findRootSvelteDependencies, needsOptimization, SvelteDependency } from './dependencies';
 import { DepOptimizationOptions } from 'vite/src/node/optimizer/index';
 
 const knownOptions = new Set([
@@ -256,7 +255,7 @@ function buildOptimizeDepsForSvelte(
 			.filter((dep) => !disabledReinclusions.includes(dep.name) && isExcluded(dep.name))
 			.flatMap((dep) =>
 				Object.keys(dep.pkg.dependencies || {})
-					.filter((depOfDep) => !isExcluded(depOfDep))
+					.filter((depOfDep) => !isExcluded(depOfDep) && needsOptimization(dep, depOfDep))
 					.map((depOfDep) => dep.path.concat(dep.name, depOfDep).join(' > '))
 			);
 		log.debug(
@@ -266,28 +265,7 @@ function buildOptimizeDepsForSvelte(
 		include.push(...transitiveDepsToInclude);
 	}
 
-	const esbuildOptions: DepOptimizationOptions['esbuildOptions'] = {
-		plugins: [
-			{
-				name: 'vite-plugin-svelte:external-node-builtin',
-				setup(build) {
-					// externalize all node builtins during prebundling
-					// https://github.com/vitejs/vite/blob/f4f0100649220453d961b6c66531c58026885680/packages/vite/src/node/plugins/resolve.ts#L224-L241
-					build.onResolve(
-						{ filter: new RegExp(`^(${builtinModules.join('|')})$`) },
-						({ path: id }) => {
-							return {
-								path: id,
-								external: true
-							};
-						}
-					);
-				}
-			}
-		]
-	};
-
-	return { include, exclude, esbuildOptions };
+	return { include, exclude };
 }
 
 function buildSSROptionsForSvelte(

